### PR TITLE
Adds Graal hints for the shaded dependencies

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
we've noticed that the shaded jctools do not work with Graal out of the box. When trying to run the native image we get the following exception

```
Caused by: java.lang.RuntimeException: java.lang.NoSuchFieldException: producerIndex
        at io.opentelemetry.internal.shaded.jctools.util.UnsafeAccess.fieldOffset(UnsafeAccess.java:111) ~[na:na]
        at io.opentelemetry.internal.shaded.jctools.queues.MpscArrayQueueProducerIndexField.<clinit>(MpscArrayQueue.java:48) ~[demo-aot-native:1.18.0]
        ... 203 common frames omitted
Caused by: java.lang.NoSuchFieldException: producerIndex
        at java.lang.Class.getDeclaredField(DynamicHub.java:968) ~[demo-aot-native:na]
        at io.opentelemetry.internal.shaded.jctools.util.UnsafeAccess.fieldOffset(UnsafeAccess.java:107) ~[na:na]
        ... 204 common frames omitted
```

That's because the shaded dependencies are missing the `META-INF/native-image/reflect-config.json`. I followed the path taken by rsocket (https://github.com/rsocket/rsocket-java/commit/f15c14a06bd5b36db532e059c678f02aa05d08be).

cc @sdeleuze

### CODE REVIEW

**Comment by sdeleuze** [comment]
@marcingrzejszczak Could you please update the PR and test it with the file moved to `sdk/trace-shaded-deps/src/main/resources/META-INF/native-image/io.opentelemetry/opentelemetry-sdk-trace/reflect-config.json` ?

**Review by trask — COMMENTED** [review]
thx @marcingrzejszczak! is there any strategy for automating this / keeping it up-to-date over time? if there's no way to automate, it would be great if you could add a test for it, so that we don't accidentally introduce regressions later on

**Comment by sdeleuze** [comment]
It is possible to run JUnit test on native via https://github.com/graalvm/native-build-tools. You can just reuse existing tests or write a simple integration one that runs on native. It requires a JVM distribution with `native-image` like [GraalVM CE](https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-22.2.0) (available on GitHub actions) or [Liberica NIK](https://bell-sw.com/liberica-native-image-kit/).

**Comment by marcingrzejszczak** [comment]
I've added tests and a Github Action build

**Comment by marcingrzejszczak** [comment]
I've added tests and an additional github action check for graal

**Comment by trask** [comment]
@marcingrzejszczak can you try `./gradlew :integration-tests:graal:spotlessApply` to fix CI spotless failure?

**Comment by marcingrzejszczak** [comment]
Is there anything else to be done here to merge this?

File: integration-tests/graal/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+plugins {
+  id("otel.java-conventions")
+  id("java-library")
```
Comment (jack-berg) [inline integration-tests/graal/build.gradle.kts:3]: No need for this - its covered in `otel.java-conventions`.

File: integration-tests/graal/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  id("org.graalvm.buildtools.native") version "0.9.14"
```
Comment (jack-berg) [inline integration-tests/graal/build.gradle.kts:4]: We centrally manage plugin versions in this repo in the `pluginManagement` block of `settings.gradle.kts`. Can you move this id / version there and omit the version here?

```
pluginManagement {
  plugins {
    ....
    id("org.graalvm.buildtools.native") version "0.9.14"
  }
}
```

File: integration-tests/graal/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    named("main") {
+      verbose.set(true)
+    }
+  }
+  toolchainDetection.set(false)
+}
+
+tasks.test {
+  useJUnitPlatform()
+}
```
Comment (jack-berg) [inline integration-tests/graal/build.gradle.kts:34]: No need for this test task config - its covered in `otel.java-conventions`.

File: integration-tests/graal/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  main {
+    val traceShadedDeps = project(":sdk:trace-shaded-deps")
+    output.dir(traceShadedDeps.file("build/extracted/shadow"), "builtBy" to ":sdk:trace-shaded-deps:extractShadowJar")
+  }
+}
+
+dependencies {
+  implementation(project(path = ":sdk:trace-shaded-deps"))
+
+  testImplementation("org.junit.jupiter:junit-jupiter")
```
Comment (jack-berg) [inline integration-tests/graal/build.gradle.kts:20]: No need for this dependency.

File: integration-tests/graal/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  id("org.graalvm.buildtools.native") version "0.9.14"
+}
+
+description = "OpenTelemetry Graal Integration Tests"
+otelJava.moduleName.set("io.opentelemetry.graal.integration.tests")
+
+sourceSets {
+  main {
+    val traceShadedDeps = project(":sdk:trace-shaded-deps")
+    output.dir(traceShadedDeps.file("build/extracted/shadow"), "builtBy" to ":sdk:trace-shaded-deps:extractShadowJar")
```
Comment (jack-berg) [inline integration-tests/graal/build.gradle.kts:13]: Mind adding a comment explaining why this is necessary?

**Review by jack-berg — APPROVED** [review]
Couple of small comments, but looks good. Thanks!

**Comment by fniephaus** [comment]
>we've noticed that the shaded jctools do not work with Graal out of the box

Does this mean that `opentelemetry-java` now works with GraalVM Native Image out of the box or is something still not working?

**Comment by marcingrzejszczak** [comment]
Since this got merged that means that indeed it does work with Native. You can check my post on Spring.io where I demonstrate that - https://spring.io/blog/2022/10/12/observability-with-spring-boot-3#running-it-all-together-with-aot-support

**Pull request merged**